src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -172,74 +171,63 @@ private static object[] ParseArguments(string methodIdentifier, string argsJson,
+            }
 
-                    if (parameterType == typeof(JSAsyncCallResult))
-                    {
-                        // We will pass the JsonDocument instance to JAsyncCallResult and make JSRuntimeBase
-                        // responsible for disposing it.
-                        shouldDisposeJsonDocument = false;
-                        // For JS async call results, we have to defer the deserialization until
-                        // later when we know what type it's meant to be deserialized as
-                        suppliedArgs[index] = new JSAsyncCallResult(jsonDocument, item);
-                    }
-                    else if (IsIncorrectDotNetObjectRefUse(item, parameterType))
-                    {
-                        throw new InvalidOperationException($"In call to '{methodIdentifier}', parameter of type '{parameterType.Name}' at index {(index + 1)} must be declared as type 'DotNetObjectRef<{parameterType.Name}>' to receive the incoming value.");
-                    }
-                    else
-                    {
-                        suppliedArgs[index] = JsonSerializer.Deserialize(item.GetRawText(), parameterType, JsonSerializerOptionsProvider.Options);
-                    }
+            // Reset the reader
+            reader = new Utf8JsonReader(utf8JsonBytes);
+            reader.Read();
 
-                    index++;
-                }
+            var suppliedArgs = new object[parameterTypes.Length];
 
-                if (shouldDisposeJsonDocument)
+            for (var i = 0; i < parameterTypes.Length; i++)
+            {
+                var parameterType = parameterTypes[i];
+                reader.Read();
+                var bytesConsumed = (int)reader.BytesConsumed;
+                if (reader.TokenType == JsonTokenType.StartObject && IsIncorrectDotNetObjectRefUse(parameterType, utf8JsonBytes.Slice(bytesConsumed), reader.CurrentState))
                 {
-                    jsonDocument.Dispose();
+                    throw new InvalidOperationException($"In call to '{methodIdentifier}', parameter of type '{parameterType.Name}' at index {(i + 1)} must be declared as type 'DotNetObjectRef<{parameterType.Name}>' to receive the incoming value.");
                 }
 
-                return suppliedArgs;
-            }
-            catch
-            {
-                // Always dispose the JsonDocument in case of an error.
-                jsonDocument?.Dispose();
-                throw;
+                suppliedArgs[i] = JsonSerializer.Deserialize(ref reader, parameterType, JsonSerializerOptionsProvider.Options);
             }
 
-            static bool IsIncorrectDotNetObjectRefUse(JsonElement item, Type parameterType)
+            return suppliedArgs;
+
+            static bool IsIncorrectDotNetObjectRefUse(Type parameterType, Span<byte> utf8JsonBytes, JsonReaderState readerState)
```

Comment:
Why not pass the reader directly rather than passing in the data/state? If you don't want the mutations to the reader be retained, pass by value (rather than ref).

Reply:
Also it might be nice to make this *not* be a local method to make it super clear that we're not expecting to capture any locals in a closure, since that would be a bad perf situation.